### ISSUE TITLE
Replace explicit gradient functions by autograd tools

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#1060]
### Subject of the issue
We have some explicit gradient functions for gaussian distribution https://github.com/pgmpy/pgmpy/blob/dev/pgmpy/sampling/base.py#L97. If we replace this with one of the auto grad tools (pytorch or tensorflow) it would allow us to extend our sampling algorithms to work on arbitrary distributions and also run on GPUs. 

### Your environment
* pgmpy version: 0.1.7
* Python version: All
* Operating System: All

### Expected behaviour
Remove all the gradient computation functions and use pytorch's autograd to do all the gradient computation. Should be able to run on GPUs as well.

**Comment by khalibartan** [comment]
@ankurankan Lemme try this with pytorch. I'm thinking we should have alternate computation backends like pymc.

**Comment by ankurankan** [comment]
@khalibartan Yeah, I would also suggest pytorch because the API is much cleaner. Also, I am already implementing Structural Equation Models using pytorch. Not sure about pymc because AFAIK it relies on theano which is no longer maintained.

**Comment by yashu-seth** [comment]
Hello everyone. Hope you are doing well. Where are we with this currently?

**Comment by ankurankan** [comment]
@yashu-seth @khalibartan So, I was looking at our code for the sampling methods and seems like we are running HMC (and NUTS) on joint distributions instead of factorized distribution (according to Bayesian Networks). If we want to run these algorithms on full distributions, a better option would be to convert our models into pymc models and use their implementations. 

Also, I can't find any specific papers on applying HMC in case of Bayesian Networks (send me links if you know any). But I think the original algorithm can be easily modified to work with Bayesian Networks. If we look at Algorithm 1 in [1], we can easily use the Bayesian Network to compute $ L(\theta) $ instead of using the full distribution but I am not sure how much performance gain we would achieve doing that. Thoughts?

[1] https://arxiv.org/pdf/1111.4246v1.pdf

**Comment by khalibartan** [comment]
@ankurankan When I was working on HMC and NUTS I couldn't find any papers of where these were being used in case of Bayesian Networks. But yeah, mathematically we should be able to gain performace improvement if we use bayesian networks to compute the log and gradient instead of using the full distribution. I believe performance gain would be noticeable difference. I'm not sure how we can use factors directly (Suggestions?).

**Comment by ankurankan** [comment]
@khalibartan I am not so sure that we will have much performance improvements in the general case because we will be evaluating the same distribution just as a product of smaller distributions. We will definitely save some memory though because we will require fewer parameters to represent the same distributions and therefore fewer gradient values also. But I don't think it's a significant gain.

But in cases when we have variables independent from other variables (through d-separation) we might have significant improvements because we won't need to consider the independent variables while sampling which should result in a smaller dimensional space to sample over.

In both these cases, I think the best solution would be to do some preprocessing on our side (dealing with independent variables etc) and then convert the preprocessed model into a pymc3 model and use their implementation. Thoughts?
